pyqtgraph/palette.py

```diff
@@ -0,0 +1,414 @@
+        **Arguments**
+        cmap           a ColorMap object to be sampled. If not given, a default color map is used
+        n_colors       default '8': Number of assigned colors. 
+                       The default set needs to include 'p0' to 'p7'
+        prefix         default 'p' assigns colors as 'p0' to 'pXX', at least 'p7' is required.
+                       Additional sets can be defined with a different prefix, e.g. 'col_0' to 'col_7'
+                       All prefixes need to start with 'p' or 'col' to avoid namespace overlap with 
+                       functional colors and hexadecimal numbers
+        start          first sampled value (default is 0.000)
+        step           step between samples. Default 'None' equally samples n colors from a 
+                       linear colormap, including values 0.0 and 1.0.
+                       Color values > 1. and < 0. wrap around!
+        =============  ===============================================================================
+        """
+        if not (
+            prefix[0]=='p' or
+            prefix[0]=='m' or
+            ( len(prefix)>=3 and prefix[:3]=='col')
+        ):
+            raise ValueError("'prefix' of plot color needs to start with 'p', 'm' or 'col'.")
+        if cmap is None: 
+            cmap = self.cmap
+        if isinstance(cmap, str):
+            cmap = colormap.get(cmap) # obtain ColorMap if identifier is given
+        if cmap is None:
+            raise ValueError("Please specify 'cmap' parameter when no default colormap is available.")
+        if not isinstance( cmap, colormap.ColorMap ):
+            raise ValueError("Failed to obtain ColorMap object for 'cmap' = '+str(cmap).")
+            
+        if prefix == 'p':
+            self.cmap = cmap # replace default color map
+            n_colors = 8 # always define 8 primary plot colors
+        if step is None:
+            step = 1 / (n_colors - 1) # sample 0. to 1. (inclusive) by default
+        for cnt in range(n_colors):
+            val = start + cnt * step
+            # print( val )
+            if val > 1.0 or val < 0.0: # don't touch 1.0 value
+                val = val % 1. # but otherwise map to 0 to 1 range
+            qcol = cmap[val]
+            key = prefix + str(cnt)
+            self.palette[key] = qcol
+            
+    def add(self, colors):
+        """
+        Add colors given in dictionary 'colors' to the palette
+        All colors will be converted to QColor.
+        Setting 'gr_bg' with a mean color value < 127 will set the palette's 'dark' property
+        """
+        for key in colors:
+            col = identifier_to_QColor( colors[key], color_dict=self.palette )
+            if key == 'gr_bg':
+                bg_tuple = col.getRgb()
+                self.dark = bool( sum( bg_tuple[:3] ) < 3 * 127 ) # dark mode?
+            self.palette[key] = col
+            
+    def addDefaultRamp(self):
```

Comment:
should there be something in this function, or should it be removed?

Reply:
For now, all custom gray colors are defined manually, but this function should replace that eventually.
Sorry, I forgot that this was left as a stub for now.